pkg/gce-pd-csi-driver/controller.go

```diff
@@ -431,14 +432,118 @@ func (gceCS *GCEControllerServer) ControllerGetCapabilities(ctx context.Context,
 }
 
 func (gceCS *GCEControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	glog.V(4).Infof("CreateSnapshot called with request %v", *req)
+
+	// Validate arguments
+	volumeID := req.GetSourceVolumeId()
+
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "CreateSnapshot Source Volume ID must be provided")
+	}
+	volKey, err := common.VolumeIDToKey(volumeID)
+	if err != nil {
+		return nil, status.Error(codes.NotFound, fmt.Sprintf("Could not find volume with ID %v: %v", volumeID, err))
+	}
+	if len(req.Name) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "CreateSnapshot Snapshot name must be provided")
+	}
+	snapshot, err := gceCS.CloudProvider.CreateSnapshot(ctx, volKey, req.Name)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unknown create snapshot error: %v", err))
+	}
+	t, err := time.Parse(time.RFC3339, snapshot.CreationTimestamp)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Failed to covert creation timestamp: %v", err))
+	}
+	createResp := &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      	common.GbToBytes(snapshot.DiskSizeGb),
+			Id:                 cleanSelfLink(snapshot.SelfLink),
```

Comment:
just making sure `cleanSelfLink` works properly here, since maybe snapshot selfLinks have different prefixes

Reply:
for now, snapshot has the same link format as volume